# KADAI 2026 - Nigeria Food Market Intelligence 
## 01.Data Audit

This notebook audits and prepares historical Nigerian food-price data 
for downstream analysis of food price trends, regional differences, and commodity-level pricing

## The audit focuses on:
- Understand the structure and courage of the soure datasets
- Checking data types, missing values and duplicate records
- Standardizing price units for comparability
- Validating relationships between food prices, markets, and commodities
- Producing a clean analytical dataset for further analysis
- Identify and documenting potential data-quality issues

## Load Data

In [6]:
import pandas as pd

In [7]:
food_prices = pd.read_csv("food_prices.csv")
markets = pd.read_csv("markets.csv")
commodities = pd.read_csv("commodoties (5).csv")

## Dataset Overview

In [8]:
food_prices.info()

<class 'pandas.DataFrame'>
RangeIndex: 75053 entries, 0 to 75052
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   date              75053 non-null  str    
 1   market_id         75053 non-null  int64  
 2   commodity_id      75053 non-null  int64  
 3   unit              75053 non-null  str    
 4   priceflag         75053 non-null  str    
 5   pricetype         75053 non-null  str    
 6   price             75053 non-null  float64
 7   usdprice          75053 non-null  float64
 8   normalized_unit   75053 non-null  str    
 9   normalized_price  75053 non-null  float64
dtypes: float64(3), int64(2), str(5)
memory usage: 5.7 MB


## Data Types
## Date Coverage

In [9]:
food_prices["date"] = pd.to_datetime(food_prices["date"])

In [10]:
markets.info()

<class 'pandas.DataFrame'>
RangeIndex: 117 entries, 0 to 116
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   market_id  117 non-null    int64  
 1   market     117 non-null    str    
 2   admin1     117 non-null    str    
 3   admin2     117 non-null    str    
 4   latitude   117 non-null    float64
 5   longitude  117 non-null    float64
dtypes: float64(2), int64(1), str(3)
memory usage: 5.6 KB


In [11]:
commodities.info()

<class 'pandas.DataFrame'>
RangeIndex: 41 entries, 0 to 40
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   commodity_id  41 non-null     int64
 1   commodity     41 non-null     str  
 2   category      41 non-null     str  
dtypes: int64(1), str(2)
memory usage: 1.1 KB


In [12]:
print("FOOD PRICES")
print(food_prices.columns)

FOOD PRICES
Index(['date', 'market_id', 'commodity_id', 'unit', 'priceflag', 'pricetype',
       'price', 'usdprice', 'normalized_unit', 'normalized_price'],
      dtype='str')


In [13]:
print("\nMARKETS")
print(markets.columns)


MARKETS
Index(['market_id', 'market', 'admin1', 'admin2', 'latitude', 'longitude'], dtype='str')


In [14]:
print("\nCOMMODITIES")
print(commodities.columns)


COMMODITIES
Index(['commodity_id', 'commodity', 'category'], dtype='str')


## Data Quality Checks

## Missing Values

In [15]:
food_prices.isnull().sum()

date                0
market_id           0
commodity_id        0
unit                0
priceflag           0
pricetype           0
price               0
usdprice            0
normalized_unit     0
normalized_price    0
dtype: int64

In [16]:
markets.isnull().sum()

market_id    0
market       0
admin1       0
admin2       0
latitude     0
longitude    0
dtype: int64

In [17]:
commodities.isnull().sum()

commodity_id    0
commodity       0
category        0
dtype: int64

## Duplicate Records

In [18]:
food_prices.duplicated().sum()

np.int64(0)

In [19]:
markets.duplicated().sum()

np.int64(0)

In [20]:
commodities.duplicated().sum()

np.int64(0)

In [21]:
food_prices.head()

,date,market_id,commodity_id,unit,priceflag,pricetype,price,usdprice,normalized_unit,normalized_price
0,2016-01-15,1967,55,Unit,actual,Retail,247.52,1.25,Unit,247.52
1,2016-01-15,1967,64,2.8 KG,actual,Retail,576.00,2.90,KG,205.71
2,2016-01-15,1967,71,2.7 KG,actual,Retail,414.00,2.08,KG,153.33
3,2016-01-15,1967,99,2.5 KG,actual,Retail,292.00,1.47,KG,116.80
4,2016-01-15,1967,284,L,actual,Retail,160.00,0.81,L,160.00


In [22]:
markets.head()

,market_id,market,admin1,admin2,latitude,longitude
0,1038,Jibia (CBM),Katsina,Jibia,13.08,7.24
1,1039,Mai Adoua (CBM),Katsina,Mai'Adua,13.18,8.23
2,1040,Damassack (CBM),Borno,Mobbar,13.01,12.56
3,1041,Mai Gatari (CBM),Jigawa,Maigatari,12.78,9.44
4,1967,Aba,Abia,Oboma Ngwa,5.15,7.36


In [23]:
commodities.head()

,commodity_id,commodity,category
0,55,Bread,cereals and tubers
1,64,Rice (imported),cereals and tubers
2,71,Rice (local),cereals and tubers
3,99,Yam,cereals and tubers
4,284,Fuel (diesel),non-food


In [24]:
food_prices.tail()

,date,market_id,commodity_id,unit,priceflag,pricetype,price,usdprice,normalized_unit,normalized_price
75048,2026-07-15,8361,404,300 G,aggregate,Retail,200.0,0.1500,G,0.67
75049,2026-07-15,8361,114,0.5 KG,aggregate,Retail,500.0,0.3600,KG,1000.00
75050,2026-07-15,8361,716,2.1 KG,aggregate,Retail,2000.0,1.4500,KG,952.38
75051,2026-07-15,3192,285,L,aggregate,Retail,2000.0,1.4500,L,2000.00
75052,2026-07-15,3192,254,1.3 KG,aggregate,Retail,5.0,0.0036,KG,3.85


In [25]:
markets.tail()

,market_id,market,admin1,admin2,latitude,longitude
112,10714,Nangere,Yobe,Nangere,11.86,11.07
113,10900,Jimeta,Adamawa,Demsa,9.28,12.46
114,10901,Girei,Adamawa,Gombi,9.43,12.48
115,10902,Yola,Adamawa,Demsa,9.25,12.44
116,10943,Ngelzarma,Yobe,Fune,11.69,11.62


In [26]:
commodities.tail()

,commodity_id,commodity,category
36,218,Cowpeas,pulses and nuts
37,114,Tomatoes,vegetables and fruits
38,185,Salt,miscellaneous food
39,97,Sugar,miscellaneous food
40,1043,Seasoning (maggi cube),miscellaneous food


In [27]:
print("Earliest date:", food_prices["date"].min())

Earliest date: 2016-01-15 00:00:00


In [28]:
print("Latest date:", food_prices["date"].max())

Latest date: 2026-07-15 00:00:00


In [29]:
print("Unique markets:", food_prices["market_id"].nunique())

Unique markets: 68


In [30]:
print("Unique commodities:", food_prices["commodity_id"].nunique())

Unique commodities: 41


In [31]:
food_prices["price"].describe()

count     75053.000000
mean       4990.140548
std       10232.171543
min           1.000000
25%         275.000000
50%         900.000000
75%        3233.330000
max      138000.000000
Name: price, dtype: float64

In [32]:
food_prices["normalized_price"].describe()

count    75053.000000
mean       523.193933
std        891.113828
min          0.020000
25%        122.900000
50%        246.290000
75%        541.820000
max      19375.000000
Name: normalized_price, dtype: float64

In [33]:
missing_markets = food_prices.loc[
   ~food_prices["market_id"].isin(markets["market_id"]),
   "market_id"
].unique()
missing_commodities = food_prices.loc[
    ~food_prices["commodity_id"].isin(commodities["commodity_id"]),
    "commodity_id"
].unique()
print("Unmatched market IDs:",missing_markets)
print("Unmatched commodity IDs:", missing_commodities)


Unmatched market IDs: [1037]
Unmatched commodity IDs: []


### Data Quality Issue: Unmatched Market ID

One market ID (`1037`) appears in the food-price dataset but is not present in the market dimension table.

A total of **53 observations** are associated with this unmatched market ID.

Because these records cannot be reliably linked to a market name or geographic information, they were excluded from the primary analytical dataset.

This ensures that the final dataset can be reliably joined with the available market metadata.

In [34]:
orphan_records = food_prices[~food_prices["market_id"].isin(markets["market_id"])].copy()

"""
Saving the 53 observations associated with market ID 1037 lacking a corresponding
market_dimension record
"""

'\nSaving the 53 observations associated with market ID 1037 lacking a corresponding\nmarket_dimension record\n'

In [35]:
food_prices = food_prices[food_prices["market_id"].isin(markets["market_id"])].copy()

In [36]:
food_prices[food_prices["market_id"] == 1037].head()

,date,market_id,commodity_id,unit,priceflag,pricetype,price,usdprice,normalized_unit,normalized_price


In [37]:
food_prices[food_prices["market_id"] == 1037].shape

(0, 10)

In [38]:
print("Price types:")
print(food_prices["pricetype"].value_counts())

print("\nPrice flags:")
print(food_prices["priceflag"].value_counts())

print("\nNormalized units:")
print(food_prices["normalized_unit"].value_counts())

Price types:
pricetype
Retail       60985
Wholesale    14015
Name: count, dtype: int64

Price flags:
priceflag
actual              41237
aggregate           32646
actual,aggregate     1117
Name: count, dtype: int64

Normalized units:
normalized_unit
KG        55215
G          9678
L          6593
pcs        2019
Unit       1061
Tubers      434
Name: count, dtype: int64


### Standardizing Price Units

Some observations are recorded on a gram-based unit.

To make prices comparable across observations, gram-based prices were converted to an equivalent price per kilogram by multiplying the normalized price by **1,000**.

This standardization allows the KG-based observations to be analyzed on a consistent unit basis.

In [39]:
# Converting grams to kilograms so normalized_price is comparable
mask_grams = food_prices["normalized_unit"] == "G"

food_prices.loc[mask_grams, "normalized_price"] = food_prices.loc[mask_grams, "normalized_price"] * 1000
food_prices.loc[mask_grams, "normalized_unit"] = "KG"

In [40]:
print(food_prices["normalized_unit"].value_counts())

normalized_unit
KG        64893
L          6593
pcs        2019
Unit       1061
Tubers      434
Name: count, dtype: int64


In [41]:
print("Non-positive raw prices:",(food_prices["price"] <= 0).sum())
print("Non-positive normalised prices:",(food_prices["normalized_price"] <= 0).sum())

Non-positive raw prices: 0
Non-positive normalised prices: 0


In [42]:
print((food_prices["market_id"] == 1037).sum())

0


In [43]:
# Primary analysis dataset - KG based commodities only
df_main = food_prices[food_prices["normalized_unit"] == "KG"].copy()

# Optional secondary dataset - liquids
df_liquids = food_prices[food_prices["normalized_unit"] == "L"].copy()

## Data Audit Summary

## Audit Check | Result
  
   Food-price records | 75,053
   Markets in market dataset | 117
   Commodities | 41
   Missing values | 0
   Duplicate rows | 0
   Unmatched market IDs | 1
   Orphan observations removed | 53
   Unmatched commodity IDs | 0
   Final analytical records | 64,893

## 9. Key Findings

- The food-price dataset contains 75,053 observations spanning January 2016 to July 2026.
- No missing values or duplicate rows were identified in the source datasets.
- One market ID (`1037`) was present in the food-price data but missing from the market dimension table, affecting 53 observations.
- No unmatched commodity IDs were identified.
- No non-positive raw or normalized prices were found.
- Gram-based observations were standardized to kilograms for comparability.
- The final analytical dataset contains 64,893 KG-based observations joined with market and commodity dimensions.

In [44]:
print(len(df_main), len(df_liquids))

64893 6593


In [45]:
# Merging the three tables together
df = df_main.merge(markets, on="market_id", how="left") \
    .merge(commodities, on="commodity_id", how="left")
    

In [46]:
print(len(df_main)), len(df) # Confirming..

64893


(None, 64893)

In [47]:
df.head()

,date,market_id,commodity_id,unit,priceflag,pricetype,price,usdprice,normalized_unit,normalized_price,market,admin1,admin2,latitude,longitude,commodity,category
0,2016-01-15,1967,64,2.8 KG,actual,Retail,576.0,2.90,KG,205.71,Aba,Abia,Oboma Ngwa,5.15,7.36,Rice (imported),cereals and tubers
1,2016-01-15,1967,71,2.7 KG,actual,Retail,414.0,2.08,KG,153.33,Aba,Abia,Oboma Ngwa,5.15,7.36,Rice (local),cereals and tubers
2,2016-01-15,1967,99,2.5 KG,actual,Retail,292.0,1.47,KG,116.80,Aba,Abia,Oboma Ngwa,5.15,7.36,Yam,cereals and tubers
3,2016-01-15,1980,482,100 KG,actual,Wholesale,15450.0,77.79,KG,154.50,Mubi,Adamawa,Hong,10.27,13.26,"Cassava meal (gari, yellow)",cereals and tubers
4,2016-01-15,1980,483,100 KG,actual,Wholesale,13819.0,69.58,KG,138.19,Mubi,Adamawa,Hong,10.27,13.26,Gari (white),cereals and tubers


In [48]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 64893 entries, 0 to 64892
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date              64893 non-null  datetime64[us]
 1   market_id         64893 non-null  int64         
 2   commodity_id      64893 non-null  int64         
 3   unit              64893 non-null  str           
 4   priceflag         64893 non-null  str           
 5   pricetype         64893 non-null  str           
 6   price             64893 non-null  float64       
 7   usdprice          64893 non-null  float64       
 8   normalized_unit   64893 non-null  str           
 9   normalized_price  64893 non-null  float64       
 10  market            64893 non-null  str           
 11  admin1            64893 non-null  str           
 12  admin2            64893 non-null  str           
 13  latitude          64893 non-null  float64       
 14  longitude         64893 non-null 

## Export Clean Dataset

In [50]:
df.to_excel("cleaned_food_prices.xlsx",  index=False, sheet_name="Data")